core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -263,9 +263,15 @@ private[spark] class ExecutorAllocationManager(
    */
   private def maxNumExecutorsNeeded(): Int = {
     val numRunningOrPendingTasks = listener.totalPendingTasks + listener.totalRunningTasks
-    math.ceil(numRunningOrPendingTasks * executorAllocationRatio /
-              tasksPerExecutorForFullParallelism)
-      .toInt
+    val maxNeeded = math.ceil(numRunningOrPendingTasks * executorAllocationRatio /
+      tasksPerExecutorForFullParallelism).toInt
+    if (listener.pendingSpeculativeTasks > 0 && tasksPerExecutorForFullParallelism > 1) {
+      // If we have pending speculative tasks, allocate one more executor to satisfy the
+      // locality requirements of speculative tasks
+      maxNeeded + 1
```

Comment:
This could lead to wast of executors ?

Reply:
As specified in the comments, this is to satisfy the locality requirements of speculative tasks. Let's say we have 1 normal task and 1 speculative task, in this case we should allocate 2 executors instead of 1.

Reply:
I typoed my example should be 1000 tasks with 2 tasks per executor, would results in 500 executors, we don't need to add one more

Reply:
Let's use your 1000 tasks (with 2 tasks/executor) example. The dynamic allocation scheduler currently doesn't track the underlying distribution of tasks on different executors. So it's totally possible that we have 999 tasks running, then on one executor we will have a single task running. And we have a pending speculative task for the same task index as this single task, then the speculative task can't be launched in this case. Doing a "+1" is a very simple way to address this situation.

Reply:
right but you can't guarantee that the +1 will get your speculative task or that the new executor won't end up on the same host. In our example case of all 999 already running it would give it a better chance if the 1 that finished happened to be on the same executor but what are the odds only 1 finished and it finished exactly on the same host?  Generally the reason you speculate is that you are assuming something on that host is wrong or slow so you want to run it somewhere else.  If that is the case that executor is likely to stay loaded up while other ones finish. If you have 500 executors already and your speculation configs are sane, the odds of another task finishing on a different executor are pretty good. If you don't have all your regular tasks finished, they could easily be put on the +1 host as well before your speculative has a chance to be scheduled on it and you could end up with only the 1 your counter part is on.  Since we don't have them integrated well its all just a guess/heuristic.  

I'm ok with leaving this logic to +1 like you have it (with expanded comment) since that is what it was but at the same time it can be a huge waste or resources. If every application has speculative on and you have 1000 apps running, you could easily be wasting 1000's of containers when you already have enough resources to run within.
We could definitely come up with better heuristics that don't waste so much but can be done separately.
Also can you add in a debug log statement here when we do the +1

Reply:
Thanks! You have made excellent points: the +1 gives it a better chance but not a guarantee, and the slow hosts are more likely to be loaded up. I'm OK with doing a +1 only when maxNeeded = 1, will address it.

Reply:
another option would be to do the logic for speculative separately instead of lumping it with the normal tasks.  This might be an in between the 2 options. You could for instance calculate how many you need for regular tasks, then calculate separately how many executors for speculative, if  speculative doesn't add another executor on top of the normal tasks then add one.  That would help with us double counting.
I'm fine with any of these as we can follow up to make it smarter in the future.